arangod/Pregel/Actor/include/Actor/ActorList.h

```diff
@@ -0,0 +1,145 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Markus Pfeiffer
+/// @author Julia Volmer
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include "Actor/ActorBase.h"
+#include "Actor/ActorPID.h"
+#include "Basics/Guarded.h"
+
+namespace arangodb::pregel::actor {
+
+struct ActorList {
+ private:
+  struct ActorMap : std::unordered_map<ActorID, std::shared_ptr<ActorBase>> {};
+  Guarded<ActorMap> actors;
+  struct ActorInfo {
+    ActorID id;
+    std::string_view type;
+  };
+
+ public:
+  ActorList(ActorMap map) : actors{std::move(map)} {}
+  ActorList() = default;
+
+  auto find(ActorID id) const -> std::optional<std::shared_ptr<ActorBase>> {
+    return actors.doUnderLock(
+        [id](ActorMap const& map) -> std::optional<std::shared_ptr<ActorBase>> {
+          auto a = map.find(id);
+          if (a == std::end(map)) {
+            return std::nullopt;
+          }
+          auto& [_, actor] = *a;
+          return actor;
+        });
+  }
+
+  auto add(ActorID id, std::shared_ptr<ActorBase> actor) -> void {
+    actors.doUnderLock(
+        [&id, &actor](ActorMap& map) { map.emplace(id, std::move(actor)); });
+  }
+
+  auto remove(ActorID id) -> void {
+    actors.doUnderLock([id](ActorMap& map) { map.erase(id); });
+  }
+
+  auto removeIf(
+      std::function<bool(std::shared_ptr<ActorBase> const&)> deletable)
+      -> void {
+    actors.doUnderLock([&deletable](ActorMap& map) {
```

Comment:
But I want to make sure that the function given to removeIf has the given signature: If it does not have this signature - meaning one input of ActorBase type and the return type bool - removeIf will not work. I don't see why we should make this function more general although it will not work in all other cases.